src/Drips.sol

```diff
@@ -705,6 +738,35 @@ library Drips {
         return prev.config.lt(next.config);
     }
 
+    /// @notice Calculates the amount dripped over a time range.
+    /// @param cycleSecs The cycle length in seconds.
+    /// Must be the same in all calls working on a single storage instance. Must be higher than 1.
+    /// @param amtPerSec The dripping rate
+    /// @param start The dripping start time
+    /// @param end The dripping end time
+    /// @param amt The dripped amount
+    function _drippedAmt(
+        uint256 cycleSecs,
+        uint256 amtPerSec,
+        uint256 start,
+        uint256 end
+    ) private pure returns (uint256 amt) {
+        // This function is written in Yul because it can be called thousands of times
+        // per transaction and it needs to be optimized as much as possible.
+        // The entire function costs 139 gas per execution as of Solidity 0.8.13,
+        // but when rewritten in unchecked Solidity it costed 396 gas.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            let endedCycles := sub(div(end, cycleSecs), div(start, cycleSecs))
+            let amtPerCycle := div(mul(amtPerSec, cycleSecs), AMT_PER_SEC_MULTIPLIER)
+            amt := mul(endedCycles, amtPerCycle)
+            let amtEnd := div(mul(amtPerSec, mod(end, cycleSecs)), AMT_PER_SEC_MULTIPLIER)
+            amt := add(amt, amtEnd)
+            let amtStart := div(mul(amtPerSec, mod(start, cycleSecs)), AMT_PER_SEC_MULTIPLIER)
+            amt := sub(amt, amtStart)
+        }
```

Comment:
Okay, the algorithm applies the same rounding rules as in addDeltas. That's good. 

I just wrote this fuzz test to verify my assumption
```solidity
    function testCorrectnessDrippedAmt(uint8 start, uint8 duration) public {
        warpTo(0);
        if (duration == 0) return;
        uint start = uint(start);
        uint duration = uint(duration);
        uint end = start + duration;
        uint192 fraction = uint192(uint(keccak256(abi.encode("fraction", start, duration)))%10**18);
        uint192 amtPerSec =  1.00 * 10**18 + fraction;
        DripsReceiver[] memory receivers = new DripsReceiver[](1);
        DripsConfig config = DripsConfigImpl.create(amtPerSec, uint32(start), uint32(duration));
        receivers[0] = DripsReceiver(receiver1, config);

        setDrips(defaultAsset, sender, 0, 100 * 10*18, receivers);
        warpTo(end + cycleSecs);
        uint256 dripped = _drippedAmt(cycleSecs, amtPerSec, start, end);
        receiveDrips(defaultAsset, receiver1, uint128(dripped));
    }
```

I tested it with 100k runs seems to be correct.

Reply:
So, we basically round per cycle and use this as the baseline.

Reply:
> So, we basically round per cycle and use this as the baseline.

Yes. The exact amount dripped in a specific second depends only on which second of the cycle it is. For a given `amtPerSec` the first second of every cycle will always drip the same amount. The same with `12345`th second of each cycle, each will drip the same amount. This allows safe stitching of different dripping periods, e.g. if you top-up, you only need to shift the `amtDelta` end time, you don't need to worry about the fractions. This is because in reality there are no fractions, each second only drips full units of tokens, it's just that some of them drip more and some of them less.

Reply:
I want to think it through a bit more. It seems correct what your are saying.

However I think we introduced a rounding complexity in the codebase. Which I always wanted to avoid. 

For example the rounding in `addDeltas` and `_drippedAmt` needs to happen in the same way otherwise it could lead to problems.

Reply:
I dont' like it being WET either. Unfortunately inline assembly limits any reusability. It can't define Yul functions but it can't call Solidity functions either, it's a sealed block of logic. We can't ditch assembly either, the gas savings are too big.